### PR TITLE
refactor browser message passing

### PR DESCRIPTION
**Pull request by Will-ShaoHua**
## Q
- AWS ID

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/95bfdd32-1721-40d7-b381-f3b557da67f8

- IdC

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/8cac36d5-7289-408a-a014-b059a78f7e60

- Existing connection

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/b8a6a91a-ea32-43a8-b8c0-d623a1f1cbca

- Reauthenticate

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/c05b7b53-3f67-41f9-96ae-783b3604d9da

- Signout

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/f6a05fea-616b-4518-b657-e016e73ef08a

## Toolkit
- AWS ID

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/1b4c1dbd-e89a-4198-8d09-40b984664ceb

- IdC

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/31b6bca8-7a3f-43a7-ba7b-70740cbfb33b

- Existing connection

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/cf37c6b7-7f3a-4c8a-ac04-89381c7f68ce

- Reauthenticate

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/197d369a-3af9-4c59-acd5-bc64757b0d7d

- Signout

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/f5d193e9-995f-4336-94a2-80dbb759d9d6

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactor

## Description

split of #4405 
typesafe json message sent from JCEF browser

to address tech debt comments left since 4/30 https://github.com/aws/aws-toolkit-jetbrains/pull/4265/files#r1568361295

## Checklist
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QLoginWebview.kt [hunk 1]
```diff
@@ -161,19 +158,11 @@ class QWebviewBrowser(val project: Project, private val parentDisposable: Dispos
...
-                    }
-                }
+            is BrowserMessage.Reauth -> {
+                reauth(ToolkitConnectionManager.getInstance(project).activeConnectionForFeature(QConnection.getInstance()))
             }
 
-            else -> {
-                error("received unknown command from Q browser: $command")
-            }
+            else -> {}
```
Comment (rli) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QLoginWebview.kt:165]: if the type is sealed, do we still need this else branch?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/webview/BrowserMessage.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    JsonSubTypes.Type(value = BrowserMessage.LoginBuilderId::class, name = "loginBuilderId"),
+    JsonSubTypes.Type(value = BrowserMessage.LoginIdC::class, name = "loginIdC"),
+    JsonSubTypes.Type(value = BrowserMessage.LoginIAM::class, name = "loginIAM"),
+    JsonSubTypes.Type(value = BrowserMessage.CancelLogin::class, name = "cancelLogin"),
+    JsonSubTypes.Type(value = BrowserMessage.Signout::class, name = "signout"),
+    JsonSubTypes.Type(value = BrowserMessage.Reauth::class, name = "reauth")
+)
+sealed interface BrowserMessage {
+
+    // TODO: make it data object when 232 is not supported by us as data object syntax is not available
```
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/webview/BrowserMessage.kt:31]:
```suggestion
    // FIX_WHEN_MIN_IS_233: data objects are not stable until Kotlin 1.9
    // https://kotlinlang.org/docs/whatsnew19.html#stable-data-objects-for-symmetry-with-data-classes
```

File: plugins/core/jetbrains-community/tst/software/aws/toolkits/jetbrains/core/BrowserMessageTest.kt [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+    private inline fun <reified T : Exception> assertDeserializedWillThrow(jsonStr: String) {
+        assertThatThrownBy {
+            objectMapper.readValue<BrowserMessage>(jsonStr)
+        }.isInstanceOf(T::class.java)
+    }
+
+    @BeforeEach
+    fun setup() {
+        objectMapper = ObjectMapper().registerKotlinModule()
```
Comment (rli) [inline plugins/core/jetbrains-community/tst/software/aws/toolkits/jetbrains/core/BrowserMessageTest.kt:33]:
```suggestion
        objectMapper = jacksonObjectMapper()
```
?

**Review by rli — APPROVED** [review]

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QLoginWebview.kt (hunk 1 above)
Comment (Will-ShaoHua) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QLoginWebview.kt]: nah, removed it

File: plugins/core/jetbrains-community/tst/software/aws/toolkits/jetbrains/core/BrowserMessageTest.kt (hunk 3 above)
Comment (Will-ShaoHua) [reply plugins/core/jetbrains-community/tst/software/aws/toolkits/jetbrains/core/BrowserMessageTest.kt]: donee

**Comment by Will-ShaoHua** [comment]
will upload functional test screen recording

**Pull request merged**